<a href="https://colab.research.google.com/github/blazaropinto/PDA_Data_Analysis_Python/blob/main/Clean_the_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Clean these data sets
---


The data in the following files has missing values, incorrectly formatted dates an incorrect data value, empty columns.

Open each file, inspect the data, make decisions about how to deal with incorrect or missing data and clean the data set.  

Make decisions about missing values - should the whole record be removed or is it obvious what the missing data should be?

Remove rows where you have made the decision to do so.  
Fill in values where you have made the decision to do so.  

Remove empty columns.

Format dates correctly (see [datetime worksheet](https://github.com/futureCodersSE/working-with-data/blob/main/Worksheet_functions_datetime.ipynb))    

Once you have cleaned each data set, save the data set into a CSV file on your computer.

To do this:  
```
from google.colab import files
dataframe_name.to_csv('filename.csv') 
files.download('filename.csv')
```

---

### Data set 1 - Calorie burning


In [ ]:
import pandas as pd
url1 = "https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/Data%20Cleaning%20Data%20Sets/dirtydata.csv"
url1_raw = 'https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/Data%20Cleaning%20Data%20Sets/dirtydata.csv'

dataset1 = pd.read_csv(url1_raw)
dataset1.head()

Duration          Date  Pulse  Maxpulse  Calories
0        60  '2020/12/01'    110       130     409.1
1        60  '2020/12/02'    117       145     479.0
2        60  '2020/12/03'    103       135     340.0
3        45  '2020/12/04'    109       175     282.4
4        45  '2020/12/05'    117       148     406.0

* Start by inspecting the data

In [ ]:
dataset1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32 entries, 0 to 31
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  32 non-null     int64  
 1   Date      31 non-null     object 
 2   Pulse     32 non-null     int64  
 3   Maxpulse  32 non-null     int64  
 4   Calories  30 non-null     float64
dtypes: float64(1), int64(3), object(1)
memory usage: 1.4+ KB


* The dtype of the column date is `object`, probably most of them are strings. We can use `.strptime()` on each element or just use the pandas function`.to_datetime()`

In [ ]:
dataset1.Date = pd.to_datetime(dataset1.Date)
dataset1.dtypes

Duration             int64
Date        datetime64[ns]
Pulse                int64
Maxpulse             int64
Calories           float64
dtype: object

* There are only 3 missing values, take a look at these rows

In [ ]:
dataset1[dataset1.isna().any(axis=1)]

Duration       Date  Pulse  Maxpulse  Calories
18        45 2020-12-18     90       112       NaN
22        45        NaT    100       119     282.0
28        60 2020-12-28    103       132       NaN

* The missing value in the date column can' be filled or replaced with anything, this information seems to be key for the record to mean anything so we can drop the row
* Once this is removed  fill the missing values with an appropriate statistic use `.describe()` to inspect and decide)

In [ ]:
dataset1.dropna(subset=['Date'], inplace=True)
dataset1.describe()

Duration       Pulse    Maxpulse    Calories
count   31.000000   31.000000   31.000000   29.000000
mean    69.193548  103.612903  128.806452  305.462069
std     71.064487    7.935899   13.095596   67.030470
min     30.000000   90.000000  101.000000  195.100000
25%     60.000000  100.000000  120.000000  250.700000
50%     60.000000  103.000000  128.000000  300.000000
75%     60.000000  107.000000  132.500000  345.300000
max    450.000000  130.000000  175.000000  479.000000

* The mean and median seem to be fairly similar is all the numeric values

In [ ]:
clean_dataset1 = dataset1.copy()
calories_mean = clean_dataset1.Calories.mean()
clean_dataset1.Calories = clean_dataset1.Calories.fillna(calories_mean)
display(clean_dataset1.info(), clean_dataset1.describe())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 31 entries, 0 to 31
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   Duration  31 non-null     int64         
 1   Date      31 non-null     datetime64[ns]
 2   Pulse     31 non-null     int64         
 3   Maxpulse  31 non-null     int64         
 4   Calories  31 non-null     float64       
dtypes: datetime64[ns](1), float64(1), int64(3)
memory usage: 1.5 KB


None

Duration       Pulse    Maxpulse    Calories
count   31.000000   31.000000   31.000000   31.000000
mean    69.193548  103.612903  128.806452  305.462069
std     71.064487    7.935899   13.095596   64.757586
min     30.000000   90.000000  101.000000  195.100000
25%     60.000000  100.000000  120.000000  250.700000
50%     60.000000  103.000000  128.000000  300.000000
75%     60.000000  107.000000  132.500000  342.650000
max    450.000000  130.000000  175.000000  479.000000

In [ ]:
from google.colab import files
clean_dataset1.to_csv('dataset1.csv')
files.download('dataset1.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

### Data set 2 - Presidential election voting
---



In [ ]:
import pandas as pd
url2 = "https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/Data%20Cleaning%20Data%20Sets/presDirty.csv"
url2_raw = 'https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/Data%20Cleaning%20Data%20Sets/presDirty.csv'

dataset2 = pd.read_csv(url2_raw)
dataset2.head()

cycle     branch  ...         timestamp simulations
0   2020  President  ...  03/11/2020 00:00       40000
1   2020  President  ...  02/11/2020 23:45       40000
2   2020  President  ...  02/11/2020 23:45       40000
3   2020  President  ...  31/10/2020 22:23       40000
4   2020  President  ...  30/10/2020 21:34       40000

[5 rows x 16 columns]

* Start by inspecting the data

In [ ]:
dataset2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156 entries, 0 to 155
Data columns (total 16 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   cycle                       156 non-null    int64  
 1   branch                      156 non-null    object 
 2   model                       156 non-null    object 
 3   modeldate                   156 non-null    object 
 4   candidate_inc               156 non-null    object 
 5   candidate_chal              156 non-null    object 
 6   nat_voteshare_other_hi      156 non-null    float64
 7   national_voteshare_inc_lo   156 non-null    float64
 8   national_voteshare_chal_lo  156 non-null    float64
 9   national_voteshare_3rd_lo   0 non-null      float64
 10  nat_voteshare_other_lo      156 non-null    float64
 11  national_turnout            59 non-null     float64
 12  national_turnout_hi         59 non-null     float64
 13  national_turnout_lo         59 non-

* There is an empty column that can be deleted. 
* There are another 3 columns with roughly 2/3 on its values missing. Take a look at the statistics.

In [ ]:
dataset2.drop(columns=['national_voteshare_3rd_lo'], inplace=True)
dataset2.describe()

cycle  nat_voteshare_other_hi  ...  national_turnout_lo  simulations
count   156.0              156.000000  ...         5.900000e+01        156.0
mean   2020.0                2.004223  ...         1.343265e+08      40000.0
std       0.0                0.061918  ...         5.684473e+06          0.0
min    2020.0                1.835479  ...         1.310000e+08      40000.0
25%    2020.0                1.974398  ...         1.310000e+08      40000.0
50%    2020.0                2.030170  ...         1.310000e+08      40000.0
75%    2020.0                2.048704  ...         1.330000e+08      40000.0
max    2020.0                2.085074  ...         1.475607e+08      40000.0

[8 rows x 9 columns]

--> in the 3 columns the median value is equal to its minimum value, but the mean is higher than the upper quartile. Inspect more the dataset before making a decision.

Many of this columns don't seem to have the most efficient datatype. Look at the number of different values on each columns to see if they should be treated as categories

In [ ]:
for column in dataset2.columns:
  print(column, ': ',
        len(dataset2[column].unique()),
        'different values')

cycle :  1 different values
branch :  1 different values
model :  1 different values
modeldate :  155 different values
candidate_inc :  2 different values
candidate_chal :  1 different values
nat_voteshare_other_hi :  154 different values
national_voteshare_inc_lo :  155 different values
national_voteshare_chal_lo :  155 different values
nat_voteshare_other_lo :  154 different values
national_turnout :  8 different values
national_turnout_hi :  8 different values
national_turnout_lo :  8 different values
timestamp :  136 different values
simulations :  1 different values


* All the columns with just 1 value could also be deleted as they don't provide valuable information, but keep the original dataset separated in case it's needed.
* look at the 2 values of 'candidate_inc' 


In [ ]:
clean_dataset2 = dataset2.copy()
clean_dataset2.drop(columns=['cycle', 'branch', 'model', 'candidate_chal', 'simulations'], inplace=True)

In [ ]:
clean_dataset2.candidate_inc.unique()

array(['Trump', ' '], dtype=object)

In [ ]:
clean_dataset2.candidate_inc.value_counts()

Trump    154
           2
Name: candidate_inc, dtype: int64

--> so these couple of 'blanks' are actually unrecognized missing values

* convert the 'timestamp' and the 'modeldate' columns into the correct pandas type and calculate the range of time of the data

In [ ]:
clean_dataset2.timestamp = pd.to_datetime(clean_dataset2.timestamp)
print('The data was collected from', clean_dataset2.timestamp.min(), 'to', clean_dataset2.timestamp.max())

The data was collected from 2020-01-09 19:06:00 to 2020-12-10 21:07:00


--> This suggest the missing data in candidate_inc is also 'Trump' and there is a unique value in the column so it can be also eliminated

In [ ]:
clean_dataset2[['modeldate', 'timestamp']].head()

modeldate           timestamp
0  11/03/2020 2020-03-11 00:00:00
1  11/02/2020 2020-02-11 23:45:00
2  11/02/2020 2020-02-11 23:45:00
3  10/31/2020 2020-10-31 22:23:00
4  10/30/2020 2020-10-30 21:34:00

--> 'modeldate' seems to be the data from the timestamp in US format, so it doesn't provide any information and could be deleted

In [ ]:
clean_dataset2.drop(['candidate_inc', 'modeldate'], axis=1, inplace=True)

* Get more info about the 3 columns that still have missing values

In [ ]:
clean_dataset2.sort_values('timestamp', inplace=True)

In [ ]:
clean_dataset2.groupby('national_turnout').median()

nat_voteshare_other_hi  ...  national_turnout_lo
national_turnout                          ...                     
141000000.0                     1.976343  ...          131000000.0
143000000.0                     1.903847  ...          133000000.0
146000000.0                     1.900200  ...          136000000.0
154000000.0                     1.882668  ...          143000000.0
156723120.0                     1.847821  ...          146239008.0
157972048.0                     1.842986  ...          147487920.0
158044784.0                     1.835479  ...          147560672.0

[7 rows x 6 columns]

In [ ]:
clean_dataset2.groupby('national_turnout_hi').median()

nat_voteshare_other_hi  ...  national_turnout_lo
national_turnout_hi                          ...                     
151000000.0                        1.976343  ...          131000000.0
153000000.0                        1.903847  ...          133000000.0
156000000.0                        1.900200  ...          136000000.0
164000000.0                        1.882668  ...          143000000.0
167175712.0                        1.847821  ...          146239008.0
168424640.0                        1.842986  ...          147487920.0
168497376.0                        1.835479  ...          147560672.0

[7 rows x 6 columns]

In [ ]:
clean_dataset2.groupby('national_turnout_lo').agg(['min', 'median', 'max'])

nat_voteshare_other_hi            ... national_turnout_hi             
                                       min    median  ...              median          max
national_turnout_lo                                   ...                                 
131000000.0                       1.911543  1.976343  ...         151000000.0  151000000.0
133000000.0                       1.899508  1.903847  ...         153000000.0  153000000.0
136000000.0                       1.900200  1.900200  ...         156000000.0  156000000.0
143000000.0                       1.866135  1.882668  ...         164000000.0  164000000.0
146239008.0                       1.847821  1.847821  ...         167175712.0  167175712.0
147487920.0                       1.842986  1.842986  ...         168424640.0  168424640.0
147560672.0                       1.835479  1.835479  ...         168497376.0  168497376.0

[7 rows x 18 columns]

In [ ]:
clean_dataset2[clean_dataset2.isna().any(axis=1)]

nat_voteshare_other_hi  ...           timestamp
63                 2.036718  ... 2020-01-09 19:06:00
62                 2.043273  ... 2020-02-09 21:52:00
61                 2.024914  ... 2020-03-09 21:55:00
60                 2.035542  ... 2020-04-09 22:34:00
59                 2.040212  ... 2020-05-09 22:17:00
..                      ...  ...                 ...
148                2.051627  ... 2020-10-08 11:59:00
152                2.039343  ... 2020-10-08 12:03:00
85                 2.034930  ... 2020-10-08 23:50:00
84                 2.028735  ... 2020-11-08 19:08:00
83                 2.027544  ... 2020-12-08 23:41:00

[97 rows x 8 columns]

Eliminating 97 rows from a 156-row dataframe is not an option. We've seen there are only 7 unique values on each of these 3 columns, and they all seem to be somehow related with the columns 'nat_voteshare_other_hi', 'national_voteshare_inc_lo'. To keep it simple for this exercise this will be sorted by splitting the dataframe and through replacement.

In [ ]:
clean_dataset2_part1 = clean_dataset2[clean_dataset2['nat_voteshare_other_hi'] < clean_dataset2['nat_voteshare_other_hi'].quantile(0.4)]
clean_dataset2_part1.fillna(value= {'national_turnout': 1.430000e+08, 'national_turnout_hi': 1.530000e+08, 'national_turnout_lo': 1.330000e+08}, inplace=True)

clean_dataset2_part2 = clean_dataset2[clean_dataset2['nat_voteshare_other_hi'] >= clean_dataset2['nat_voteshare_other_hi'].quantile(0.4)]
clean_dataset2_part2.fillna(value= {'national_turnout': 1.410000e+08	, 'national_turnout_hi': 1.510000e+08	, 'national_turnout_lo': 1.310000e+08	}, inplace=True)

clean_dataset2 = clean_dataset2_part1.append(clean_dataset2_part2)
clean_dataset2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 156 entries, 33 to 52
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   nat_voteshare_other_hi      156 non-null    float64       
 1   national_voteshare_inc_lo   156 non-null    float64       
 2   national_voteshare_chal_lo  156 non-null    float64       
 3   nat_voteshare_other_lo      156 non-null    float64       
 4   national_turnout            156 non-null    float64       
 5   national_turnout_hi         156 non-null    float64       
 6   national_turnout_lo         156 non-null    float64       
 7   timestamp                   156 non-null    datetime64[ns]
dtypes: datetime64[ns](1), float64(7)
memory usage: 11.0 KB


/usr/local/lib/python3.7/dist-packages/pandas/core/series.py:4536: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  downcast=downcast,
/usr/local/lib/python3.7/dist-packages/pandas/core/series.py:4536: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  downcast=downcast,


In [ ]:
from google.colab import files
clean_dataset2.to_csv('dataset2.csv') 
files.download('dataset2.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>